### PR TITLE
Feature/drawbridge mechanics

### PR DESCRIPTION
**Pull request by ArseniyShestakov**
Very basic code is ready, but proper implementation will include some rework of both movements and obstacle code. Currently I'm not planning to implement any client-side visualization, but if you have any ideas how to make future client-side programming easier please let me know.

More details about H3 drawbridge behaviour and implementation:

* All walking stacks request lowering of drawbridge from hexes before 95 and 96.
* All walking stacks request lowering from hex before 94, but only in case their destination at least hex 95. 
Exception: Fortress have moat on hex 94 so bridge will be lowered even if 94 is destination hex.
* Bridge is automatically raised when no creature standing on tiles 95 and 96. 
Exception: Fortress bridge is not raised if there creature on hex 94 as well.
* If bridge raised any stack (alive or dead) on hex 94 wont let lower it. 
Exception: destroyed gate tower make bridge constantly lowered.

### CODE REVIEW

File: lib/CBattleCallback.cpp [hunk 1]
```diff
@@ -1129,9 +1138,20 @@ AccessibilityInfo CBattleInfoCallback::getAccesibility() const
...
+		{
+		case EDrawbridgeState::RAISED:
+			accessability = EAccessibility::GATE;
+			break;
+
+		case EDrawbridgeState::RAISED_BLOCKED:
+			accessability = EAccessibility::UNAVAILABLE;
+			break;
+		}
+		ret[95] = ret[96] = accessability;
```
Comment (alexvins) [inline lib/CBattleCallback.cpp]: Use ESiegeHexes here

File: lib/GameConstants.h [hunk 2]
```diff
@@ -510,6 +510,15 @@ namespace EWallState
 	};
 }
 
+enum class EDrawbridgeState
```
Comment (alexvins) [inline lib/GameConstants.h]: Just in case define underlying type here.

File: lib/NetPacksLib.cpp [hunk 3]
```diff
@@ -1208,6 +1208,11 @@ DLL_LINKAGE void BattleObstaclePlaced::applyGs( CGameState *gs )
 	gs->curB->obstacles.push_back(obstacle);
 }
 
+DLL_LINKAGE void BattleDrawbridgeStateChanged::applyGs(CGameState *gs)
+{
+	gs->curB->si.drawbridgeState = state;
```
Comment (alexvins) [inline lib/NetPacksLib.cpp]: Should be check for curB. State change on the end of battle in theory may lead to crash.

**Comment by ArseniyShestakov** [comment]
So basic implementation is feature complete, but with two "visual" glitches related to double-wide stacks:

* When moving from outside to outer bridge hex 94 stack will stop one two one tile before bridge hex even if moving from top or down hexes. This is intended behavior when stack moving from hex 93 to 95, but not when you moving from 59 to 95.
* Bridge raising isn't handled properly for double wide stacks. So bridge will only be raised after stack movement is finished while for single-hex stacks it's raised once it's leave gate / bridge tile.

Though both problems don't affect gameplay at all while first glitch appear due to some dirty hack that fix actual game mechanics issue.

**Comment by ArseniyShestakov** [comment]
My internet is lagging like hell, so I get bored and decide to implement client visual for drawbridge.
So it's working with sound effect, but without animation. Code is bad, sorry...

**Comment by ArseniyShestakov** [comment]
So this one is imperfect and I may still want some changes like moving more hex IDs into enum all around client code. Though to actually improve drawbridge in server code I'll need to rework obstacle handling and probably movement too and I don't want to put these changes into this branch.

File: client/battle/CBattleInterface.h [hunk 4]
```diff
@@ -198,16 +198,32 @@ class CBattleInterface : public CIntObject
...
+			BACKGROUND = 0,
+			BACKGROUND_WALL = 1,
+			KEEP,
+			BOTTOM_TOWER,
+			BOTTOM_WALL,
+			WALL_BELLOW_GATE,
+			WALL_OVER_GATE,
+			UPPER_WALL,
+			UPPER_TOWER,
+			GATE = 9,
```
Comment (alexvins) [inline client/battle/CBattleInterface.h]: Is there a reason to explicitly define this?

Comment (ArseniyShestakov) [inline client/battle/CBattleInterface.h]: Yes. Enum like that is clearly much better than magic numbers with comments everywhere in code.

Comment (alexvins) [inline client/battle/CBattleInterface.h]: You misunderstood me. Why  "GATE = 9" and everything else is auto?

Comment (ArseniyShestakov) [inline client/battle/CBattleInterface.h]: Absolutely no reason, just forgot to remove it.

**Pull request merged**